### PR TITLE
Streaming SQL Query Responses with Server-Sent Events (SSE)

### LINKED ISSUE #7: Implement FastAPI SSE endpoint to stream SQL Query responses to frontend

### PR DESCRIPTION
**Pull request by MalayJoshi-Karya**
This PR introduces the implementation of a server-side FastAPI application that streams SQL query responses using Server-Sent Events (SSE). This PR also includes a client-side Python script that sends queries to the API and processes the SSE responses. The client-side script is not the final version but serves as a visualization of how the server-sse.py will interact with the frontend via SSE.

This commit closes #7. 

## **Key Features:**
- **Streaming SQL Responses:** The server streams SQL query responses as Server-Sent Events (SSE) in real-time, allowing the client to receive responses incrementally as they are generated.
- **SQL Generation:** The server uses a dummy SQL generator function, which will later be replaced by the actual ```text_to_sql``` module that converts natural language queries into SQL queries.
- **Session Handling:** The server generates a new session ID for each query if not provided, ensuring that the queries can be tracked and handled individually.
- **Logging:** Comprehensive logging has been added for both client and server to track:
-- Query processing status.
-- Errors and exceptions during execution.
-- Event streaming progress.
-- Response data at different stages.

## **Example Usage:**
- **Starting the Server:**
-- First, start the FastAPI server by running the server-side script.
-- The server listens for incoming queries on http://127.0.0.1:8000/chat.

- **Sending a Query from the Client-side:**
-- The client sends a **input request in JSON format** to the server using an asynchronous HTTP GET request:
```
{
"user_query": string, 
"session_id": [optional]string , 
"type": None/"report" or "reply"
}
```

-- In starting when a new session starts, example of input request sent to server:
```
{
"user_query": "What is the total sales for last month?", 
"session_id": None,
"type": None/"report" (if user has requested just for the by-default response of **table-based response** then type will be ```None/"report"```, but if user has requested for **summary response** then type will be ```"reply"```)
}
```

-- In consecutive multi-turn conversations within a same session, example of input request sent to server:
```
{
"user_query": some string, 
"session_id": string generated by server during start of session
"type": None/"report" (assuming user has still requested for table-based response)
}
```

- **Response Generation and Streaming from Server-side:**
-- Upon receiving input request from client-side, the server processes it by generating an SQL query (via the dummy ```text_to_sql``` module for now) and streams the **output responses in JSON format** as Server-Sent Events (SSE) asbelow, which the client receives incrementally:
```
{
  "response": {
    "type": None/"report" or "reply",
    "payload": string
  },
  "session_id": string
}
```

-- In starting when a new session starts, Server-side generate a unique ```session_id``` which is then used for all next set of multi-conversations till that session continues.

-- Upon receiving input request from client-side, if ```type``` is  ```None/"report"```, then in output response also ```type``` is  ```None/"report"```. Else, if in input request ```type``` is  ```reply```, then in output response also ```type``` is  ```reply```

- **Client-Side Processing:**
-- The client-side script handles the SSE by logging each received event and displaying the SQL query responses in real-time.
-- Client-side log (partial):
```
2024-10-15 14:23:45 - INFO - Received event: {"response": {"type": "reply", "payload": "SELECT SUM(sales) FROM sales WHERE date >= '2024-09-01' AND date <= '2024-09-30';"}, "session_id": "abc123"}
2024-10-15 14:23:50 - INFO - Received event: {"response": {"type": "reply", "payload": "Proc
```
- **Final Output:**
-- The final logged output would reflect a successful SQL query generation and streaming. Once completed, the client logs the final event and the server logs the completion of the query processing.

### CODE REVIEW

File: backend/fastapi-sse/text_to_sql.py [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+
+load_dotenv()
+my_api_key = os.getenv("GEMINI_API_KEY")
+
+if not my_api_key:
+    raise ValueError("API key not found. Make sure your .env file contains GEMINI_API_KEY.")
+print("API Key loaded successfully!")
+
+genai.configure(api_key=my_api_key)
+model = genai.GenerativeModel("gemini-1.5-flash") # gemini-1.5-pro
```
[the remark this answers is not in the document]
Comment (MalayJoshi-Karya) [reply backend/fastapi-sse/text_to_sql.py]: Thank you for the suggestion! I've moved the model name to the .env file as recommended in the latest commit.

File: backend/fastapi-sse/server-sse.py [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+error_handler = logging.FileHandler('error.log')
+error_handler.setFormatter(logging.Formatter(log_format))
+error_handler.setLevel(logging.ERROR)
+
+# Create a logger and add the handlers
+logger = logging.getLogger(__name__)
+logger.setLevel(logging.DEBUG)
+logger.addHandler(console_handler)
+logger.addHandler(success_handler)
+logger.addHandler(error_handler)
```
[the remark this answers is not in the document]
Comment (MalayJoshi-Karya) [reply backend/fastapi-sse/server-sse.py]: I've consolidated the logger setup into a single file and updated the imports accordingly in the latest commit.

File: backend/fastapi-sse/requirements.txt [hunk 3]
```diff
@@ -0,0 +1,6 @@
+pip install fastapi
+pip install uvicorn
+pip install aiohttp    
+pip install requests      
+pip install python-dotenv  
+pip install google-generativeai --quiet
```
[the remark this answers is not in the document]
Comment (MalayJoshi-Karya) [reply backend/fastapi-sse/requirements.txt]: Thank you for the input! I've updated the dependencies to specify the library versions in requirements.txt, and removed the pip install instructions as suggested.

**Comment by SoumyaRanjanPatnaik** [comment]
I tried running the code.
The server code runs fine, but `test-client-sse.py` gives the following error:


> 2024-10-18 23:50:53,420 - ERROR - AIOHTTP error occurred: Response payload is not completed: <TransferEncodingError: 400, message='Not enough data for satisfy transfer length header.'>

**Comment by MalayJoshi-Karya** [comment]
> I tried running the code. The server code runs fine, but `test-client-sse.py` gives the following error:
> 
> > 2024-10-18 23:50:53,420 - ERROR - AIOHTTP error occurred: Response payload is not completed: <TransferEncodingError: 400, message='Not enough data for satisfy transfer length header.'>

Soumya, this code is running in my system and even demoed the whole flow to Abdullah as well last week. Might be possible there is some port issue in your local. We could look up to your system specific error and fix it in your local system later in a separate meeting. 

For now you could go ahead with merging this PR with main, as ```server-sse.py``` is the main file with endpoints exposed and as it's working as intended, thus you and Abdullah could directly get response from endpoints of this file (even if ```test-client-sse.py``` is not working as intended in your system as we could look up and fix your system specific error in a separate meeting).

**Comment by SoumyaRanjanPatnaik** [comment]
I'm pretty sure this is not a system issue, but rather an environment issue (includes the .env file and maybe even the python environment). My suggestion: create a `sample.env` file in a followup pr so that it is clear what environment variables are necessary. I'll merge the PR for now

**Comment by MalayJoshi-Karya** [comment]
> I'm pretty sure this is not a system issue, but rather an environment issue (includes the .env file and maybe even the python environment). My suggestion: create a `sample.env` file in a followup pr so that it is clear what environment variables are necessary. I'll merge the PR for now

Thank you Soumya for review and moving forward to merge the PR. 

Just in case, if you think it's not system issue, and you would like to look up to required environment variables, then please check out the README file. The python version is also mentioned in requirements.txt file along with specified versions of other required libraries. Thank you once again for reviewing this PR :)

**Review by SoumyaRanjanPatnaik — COMMENTED** [review]
LGTM. Add sample.env in a seperate PR.

**Comment by MalayJoshi-Karya** [comment]
> LGTM. Add sample.env in a seperate PR.

Sure will do that Soumya, you can proceed forward to merge this PR.

**Pull request merged**